### PR TITLE
Implement new retargeting algorithm

### PR DESCRIPTION
**Pull request by raychu86**
## Motivation

This PR implements the new coinbase and proof retargeting algorithm as follows:

```
target = u64::MAX / HashToU64(cm)

delta = time_block_i - time_block_{i-1}
factor = (delta - duration_round) / # rounds per epoch
target_coinbase' = target_coinbase * 2^-factor
target_proof' = target_coinbase' >> 10
```

### CODE REVIEW

**Comment by HarukaMa** [comment]
It seems that the `cumulative_prover_target` is calculated by simply summing all calculated targets of the proofs collected so far; this could make the PoSW part luck-based: if a prover, by sheer luck, generates a proof with a very high target (say much higher than the coinbase target), this one prover would get almost all of the rewards for that block. 

Counting by proof numbers might make the coinbase rewards more expectable by provers.

File: vm/compiler/src/ledger/mod.rs [hunk 1]
```diff
@@ -76,18 +76,20 @@ const BLOCKS_DEPTH: u8 = 32;
 // TODO (raychu86): Move the following constants to a dedicated space (Or include in Network).
 
 /// The expected time per block in seconds.
-pub const ANCHOR_TIME: u64 = 15;
+pub const ANCHOR_TIME: i64 = 15;
```
Comment (howardwu) [inline vm/compiler/src/ledger/mod.rs]:
```suggestion
pub const ANCHOR_TIME: i64 = 20;
```

File: vm/compiler/src/ledger/mod.rs [hunk 2]
```diff
@@ -76,18 +76,20 @@ const BLOCKS_DEPTH: u8 = 32;
 // TODO (raychu86): Move the following constants to a dedicated space (Or include in Network).
 
 /// The expected time per block in seconds.
```
Comment (howardwu) [inline vm/compiler/src/ledger/mod.rs]:
```suggestion
/// The anchor time per block in seconds, which must be greater than the round time per block.
```

File: vm/compiler/src/ledger/block/header/metadata/genesis.rs [hunk 3]
```diff
@@ -45,7 +45,7 @@ impl<N: Network> Metadata<N> {
             // Ensure the proof target in the genesis block is `GENESIS_PROOF_TARGET`.
             && self.proof_target == GENESIS_PROOF_TARGET
             // Ensure the timestamp in the genesis block is `ANCHOR_TIMESTAMP`.
```
Comment (howardwu) [inline vm/compiler/src/ledger/block/header/metadata/genesis.rs]:
```suggestion
            // Ensure the timestamp in the genesis block is `GENESIS_TIMESTAMP`.
```

File: vm/compiler/src/ledger/helpers/mod.rs [hunk 4]
```diff
@@ -119,8 +209,11 @@ mod tests {
     const NUM_GATES_PER_CREDIT: u64 = 1_000_000; // 1 million gates == 1 credit
     const STARTING_SUPPLY: u64 = 1_000_000_000 * NUM_GATES_PER_CREDIT; // 1 quadrillion gates == 1 billion credits
 
-    const ANCHOR_TIMESTAMP: u64 = 1640179531; // 2019-01-01 00:00:00 UTC
-    const ANCHOR_TIME: u64 = 15; // 15 seconds
+    const ANCHOR_TIMESTAMP: i64 = 1640179531; // 2019-01-01 00:00:00 UTC
+    const ANCHOR_TIME: i64 = 15; // 15 seconds
```
Comment (howardwu) [inline vm/compiler/src/ledger/helpers/mod.rs]: Why aren't we using the constants from the real file?

File: vm/compiler/src/ledger/mod.rs [hunk 5]
```diff
@@ -75,19 +75,21 @@ const BLOCKS_DEPTH: u8 = 32;
...
 // TODO (raychu86): Move the following constants to a dedicated space (Or include in Network).
 
-/// The expected time per block in seconds.
-pub const ANCHOR_TIME: u64 = 15;
+/// The anchor time per block in seconds, which must be greater than the round time per block.
+pub const ANCHOR_TIME: i64 = 20;
 /// The fixed timestamp of the genesis block.
-pub const ANCHOR_TIMESTAMP: u64 = 1663718400; // 2022-09-21 00:00:00 UTC
+pub const GENESIS_TIMESTAMP: i64 = 1663718400; // 2022-09-21 00:00:00 UTC
+/// The expected number of rounds per epoch (2 hours).
```
Comment (howardwu) [inline vm/compiler/src/ledger/mod.rs]:
```suggestion
/// The number of rounds per epoch (2 hours).
```

File: vm/compiler/src/ledger/helpers/mod.rs [hunk 6]
```diff
@@ -109,6 +87,118 @@ pub(crate) fn estimated_block_height(anchor_time: u64, num_years: u32) -> u64 {
...
+///
+fn retarget<const ANCHOR_TIME: i64, const NUM_ROUNDS_PER_EPOCH: u32>(
+    previous_target: u64,
+    previous_block_timestamp: i64,
+    block_timestamp: i64,
+    is_inverse: bool,
+) -> u64 {
+    // Compute the difference in block time elapsed, defined as:
+    let mut drift = {
+        // Determine the block time elapsed (in seconds) since the anchor block.
```
Comment (howardwu) [inline vm/compiler/src/ledger/helpers/mod.rs]:
```suggestion
        // Determine the block time elapsed (in seconds) since the previous block.
```

**Comment by raychu86** [comment]
> It seems that the `cumulative_prover_target` is calculated by simply summing all calculated targets of the proofs collected so far; this could make the PoSW part luck-based: if a prover, by sheer luck, generates a proof with a very high target (say much higher than the coinbase target), this one prover would get almost all of the rewards for that block.

Thanks for the observation! This is a scenario we've acknowledged previously, and believe it is not a significant issue - we plan to have a full writeup of the protocol/game-theory soon to help explain and draw discussion to these cases.

File: vm/compiler/src/ledger/helpers/mod.rs [hunk 7]
```diff
@@ -14,281 +14,340 @@
...
+    Ok(u64::try_from(STARTING_SUPPLY as u128 * 25 / 1000 / block_height_around_year_1 as u128)?)
 }
 
-/// Calculate the proving reward for a given block.
-pub(crate) fn proving_reward<const STARTING_SUPPLY: u64, const ANCHOR_TIMESTAMP: u64, const ANCHOR_TIME: u64>(
-    num_validators: u64,
-    timestamp: u64,
+///
+/// Calculate the coinbase reward for a given block.
+///     R_coinbase = max(0, H_Y10 - H) * R_anchor * 2^((D - B) / N).
```
Comment (qy3u) [inline vm/compiler/src/ledger/helpers/mod.rs]: This comment seems conflict with the implemention and comment below. I think the last factor should be 
`2^((B - D) / N)` 
or 
`2^(-1 * ((D - B) / N))` 
here.

Comment (howardwu) [reply vm/compiler/src/ledger/helpers/mod.rs]: Thanks for catching this!

File: vm/compiler/src/ledger/latest.rs [hunk 8]
```diff
@@ -62,8 +62,41 @@ impl<N: Network, B: BlockStorage<N>, P: ProgramStorage<N>> Ledger<N, B, P> {
...
+                if self.get_header(previous_round_number)?.round() != previous_round_number as u64 {
+                    bail!("Round number does not match the block height")
+                }
+
+                block_hash
+            }
+        };
+
+        EpochChallenge::new(latest_epoch_number, epoch_block_hash, COINBASE_PUZZLE_DEGREE)
     }
```
Comment (howardwu) [inline vm/compiler/src/ledger/latest.rs]: We need to change the following:
- `self.current_round` -> `self.current_height`
- `NUM_ROUNDS_PER_EPOCH` -> `NUM_BLOCKS_PER_EPOCH`

These can be updated in a new PR.

File: vm/compiler/src/ledger/block/mod.rs [hunk 9]
```diff
@@ -160,6 +160,11 @@ impl<N: Network> Block<N> {
         self.header.round()
     }
 
+    /// Returns the epoch number of this block.
+    pub const fn epoch_number(&self) -> u64 {
+        self.round() / NUM_ROUNDS_PER_EPOCH as u64
+    }
+
```
Comment (howardwu) [inline vm/compiler/src/ledger/block/mod.rs]: Ditto.

**Comment by howardwu** [comment]
^Rebased here.

File: vm/compiler/src/ledger/latest.rs (hunk 8 above)
Comment (howardwu) [reply vm/compiler/src/ledger/latest.rs]: I've handled the changes in commit https://github.com/AleoHQ/snarkVM/pull/1127/commits/7aeb75577d8fcd59ca09c0bc712c311fcf549166 in this PR.

**Comment by howardwu** [comment]
^This commit switches epoch calculation from `round` to `height`.

**Comment by howardwu** [comment]
^This commit changes the `coinbase_reward` half-life to `ANCHOR_TIME`, and switches `ANCHOR_TIME` to a `u16`.

**Review by howardwu — APPROVED** [review]

**Pull request merged**